lib/http3/common.c

```diff
@@ -543,7 +543,7 @@ static void process_packets(h2o_quic_ctx_t *ctx, quicly_address_t *destaddr, qui
         if (iter != kh_end(ctx->conns_by_id)) {
             conn = kh_val(ctx->conns_by_id, iter);
             /* CID-based matching on Initial and 0-RTT packets should only be applied for clients */
-            if (!quicly_is_client(conn->quic) && packets[0].cid.dest.might_be_client_generated)
+            if (!quicly_is_destination(conn->quic, &destaddr->sa, &srcaddr->sa, packets))
```

Comment:
The root cause is that all Initial / 0-RTT packets that carry server-generated CIDs are dropped. The fix is to call the correct helper function.